pkg/controller/intrusiondetection/intrusiondetection_controller.go

```diff
@@ -88,134 +88,121 @@ func Add(mgr manager.Manager, opts options.AddOptions) error {
 		return err
 	}
 
-	go utils.WaitToAddLicenseKeyWatch(controller, k8sClient, log, licenseAPIReady)
+	// Make a helper for determining which namespaces to use based on tenancy mode.
+	helper := utils.NewNamespaceHelper(opts.MultiTenant, render.ManagerNamespace, "")
 
-	go utils.WaitToAddResourceWatch(controller, k8sClient, log, dpiAPIReady,
-		[]client.Object{&v3.DeepPacketInspection{TypeMeta: metav1.TypeMeta{Kind: v3.KindDeepPacketInspection}}})
+	installNS, _, _ := tenancy.GetWatchNamespaces(opts.MultiTenant, render.IntrusionDetectionNamespace)
```

Comment:
If this is in multi tenant mode, then this value seems to be empty. What does that mean in terms of the policy / job watches? Wouldn't this make it so you're technically watching across all namespaces for these policies / jobs (which would still be filtered by name).

Future improvements are going to improve our caching mechanisms to cache by the namespace for namespace resources so we don't load cache resources from other namespaces. If we don't have the specific namespace we want to watch for the resource in, we can't achieve this and we'll continue to load all resources into the cache.

Reply:
Yes, in multi-tenant mode we currently watch all namespaces. This is because we don't know the total set of namespaces that we need to watch at startup.

if / when we make that change, we will need to handle this case appropriately - either by omitting multi-tenancy from that improvement (which would be sad) or perhaps by implementing a more dynamic watch system that enables us to add / remove namespaces to be watched as tenants in the cluster come and go.

For now, this is the pattern that we're already using throughout the rest of the code.

Reply:
This is unfortunate, but I suppose I understand the difficulty.